src/utils/shallowEqual.js

```diff
@@ -23,28 +23,25 @@ function shallowEqual(objA, objB) {
     return true;
   }
 
-  if (!objA || !objB) {
+  if (typeof objA !== 'object' || objA === null
+      || typeof objB !== 'object' || objB === null) {
     return false;
   }
 
-  if (typeof objA !== 'object' || typeof objB !== 'object') {
+  var keysA = Object.keys(objA), keysB = Object.keys(objB);
+
+  if (keysA.length !== keysB.length) {
     return false;
   }
 
-  var key;
   // Test for A's keys different from B.
-  for (key in objA) {
-    if (objA.hasOwnProperty(key) &&
-        (!objB.hasOwnProperty(key) || objA[key] !== objB[key])) {
-      return false;
-    }
-  }
-  // Test for B's keys missing from A.
-  for (key in objB) {
-    if (objB.hasOwnProperty(key) && !objA.hasOwnProperty(key)) {
+  var bHasOwnProperty = objB.hasOwnProperty.bind(objB);
```

Comment:
Might want to do `{}.hasOwnProperty.bind(objB)` (not that what you have is different than what we already have, but if we're here we might as well make sure we aren't hitting cases where people override `hasOwnProperty`)